actor.Stopped being sent twice

## `actor.Stopped` message is being sent twice to an actor when:
 - actor crashes with `panic()`
 - actor has no restarts left

## How to reproduce
```go
package main

import (
	"fmt"
	"time"

	"github.com/anthdm/hollywood/actor"
)

type foo struct {
	quitch chan struct{}
}

type bar struct{}

func (f *foo) Receive(ctx *actor.Context) {
	switch ctx.Message().(type) {
	case actor.Started:
		fmt.Println("actor.Started")

	case bar:
		go func() {
			tick := time.NewTicker(time.Second * 1)
			defer tick.Stop()

			for {
				select {
				case <-f.quitch:
					fmt.Println("quit signal, closing goroutine")
					return
				case <-tick.C:
					fmt.Println("goroutine running")
				}
			}
		}()

		time.Sleep(time.Second * 5)
		panic("panic in foo.Receive -> bar")

	case actor.Stopped:
		fmt.Println("actor.Stopped")
		close(f.quitch)
	}
}

func NewFoo() actor.Producer {
	return func() actor.Receiver {
		return &foo{quitch: make(chan struct{})}
	}
}

func main() {
	engine, err := actor.NewEngine(actor.NewEngineConfig())
	if err != nil {
		panic(err)
	}

	pid := engine.Spawn(
		NewFoo(),
		"foo", actor.WithMaxRestarts(0))

	engine.Send(pid, bar{})

	time.Sleep(time.Second * 100)
}
```

## output:
```bash
actor.Started
goroutine running
goroutine running
goroutine running
goroutine running
goroutine running
actor.Stopped
quit signal, closing goroutine
actor.Stopped
panic: panic in foo.Receive -> bar [recovered]
        panic: close of closed channel
        panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x2 addr=0x0 pc=0x1020cf008]
```
Notice `actor.Stopped` appears twice. This can cause problems in some operations, such as trying to close a channel twice - which can cause unexpected errors if one was to assume that actor.Stopped can only happen once.


### Expected output
```bash
actor.Started
goroutine running
goroutine running
goroutine running
goroutine running
goroutine running
actor.Stopped
quit signal, closing goroutine
2025/04/09 14:34:17 ERROR Actor crashed too many times pid=foo/111950105150544719
```



---

Some more info:
- It only happens when the actor is on its last restart, so it has no restarts left. This can be seen more clear when you run it with `actor.WithMaxRestarts(0)`. 
- Running an actor with `actor.WithMaxRestarts(math.MaxInt)` would effectively solve this issue, as it only occurs when `p.restarts == p.MaxRestarts`

--- a/actor/context.go
+++ b/actor/context.go
@@ -17,6 +17,8 @@ type Context struct {
 	engine   *Engine
 	receiver Receiver
 	message  any
+	// function to get the current number of messages in the inbox
+	getInboxCount func() int
 	// the context of the parent if we are a child.
 	// we need this parentCtx, so we can remove the child from the parent Context
 	// when the child dies.
@@ -31,6 +33,9 @@ func newContext(ctx context.Context, e *Engine, pid *PID) *Context {
 		engine:   e,
 		pid:      pid,
 		children: safemap.New[string, *PID](),
+		getInboxCount: func() int {
+			return -1
+		},
 	}
 }
 
@@ -173,4 +178,9 @@ func (c *Context) Engine() *Engine {
 // Message returns the message that is currently being received.
 func (c *Context) Message() any {
 	return c.message
+}
+
+// GetInboxCount returns the number of messages in the inbox of the current process.
+func (c *Context) GetInboxCount() int {
+	return c.getInboxCount()
 }

--- a/actor/inbox.go
+++ b/actor/inbox.go
@@ -42,6 +42,7 @@ type Inboxer interface {
 	Send(Envelope)
 	Start(Processer)
 	Stop() error
+	Count() int
 }
 
 type Inbox struct {
@@ -108,4 +109,8 @@ func (in *Inbox) Start(proc Processer) {
 func (in *Inbox) Stop() error {
 	atomic.StoreInt32(&in.procStatus, stopped)
 	return nil
+}
+
+func (in *Inbox) Count() int {
+	return int(in.rb.Len())
 }

--- a/actor/process.go
+++ b/actor/process.go
@@ -70,9 +70,6 @@ func (p *process) Invoke(msgs []Envelope) {
 		// If we recovered, we buffer up all the messages that we could not process
 		// so we can retry them on the next restart.
 		if v := recover(); v != nil {
-			p.context.message = Stopped{}
-			p.context.receiver.Receive(p.context)
-
 			p.mbuffer = make([]Envelope, nmsg-nproc)
 			for i := 0; i < nmsg-nproc; i++ {
 				p.mbuffer[i] = msgs[i+nproc]
@@ -121,8 +118,6 @@ func (p *process) Start() {
 	p.context.receiver = recv
 	defer func() {
 		if v := recover(); v != nil {
-			p.context.message = Stopped{}
-			p.context.receiver.Receive(p.context)
 			p.tryRestart(v)
 		}
 	}()
@@ -166,6 +161,9 @@ func (p *process) tryRestart(v any) {
 		return
 	}
 
+	p.context.message = Stopped{}
+	p.context.receiver.Receive(p.context)
+
 	p.restarts++
 	// Restart the process after its restartDelay
 	p.context.engine.BroadcastEvent(ActorRestartedEvent{
@@ -180,7 +178,9 @@ func (p *process) tryRestart(v any) {
 }
 
 func (p *process) cleanup(cancel context.CancelFunc) {
-	defer cancel()
+	if cancel != nil {
+		defer cancel()
+	}
 
 	if p.context.parentCtx != nil {
 		p.context.parentCtx.children.Delete(p.pid.ID)

--- a/actor/process.go
+++ b/actor/process.go
@@ -6,6 +6,7 @@ import (
 	"fmt"
 	"log/slog"
 	"runtime/debug"
+	"sync/atomic"
 	"time"
 
 	"github.com/DataDog/gostackparse"
@@ -33,6 +34,7 @@ type process struct {
 	pid      *PID
 	restarts int32
 	mbuffer  []Envelope
+	mcount   int32
 }
 
 func newProcess(e *Engine, opts Opts) *process {
@@ -45,6 +47,7 @@ func newProcess(e *Engine, opts Opts) *process {
 		context: ctx,
 		mbuffer: nil,
 	}
+	ctx.getInboxCount = p.Count
 	return p
 }
 
@@ -66,6 +69,7 @@ func (p *process) Invoke(msgs []Envelope) {
 		// for bookkeeping.
 		processed = 0
 	)
+	atomic.StoreInt32(&p.mcount, int32(nmsg))
 	defer func() {
 		// If we recovered, we buffer up all the messages that we could not process
 		// so we can retry them on the next restart.
@@ -74,12 +78,14 @@ func (p *process) Invoke(msgs []Envelope) {
 			for i := 0; i < nmsg-nproc; i++ {
 				p.mbuffer[i] = msgs[i+nproc]
 			}
+			atomic.StoreInt32(&p.mcount, int32(nmsg-nproc))
 			p.tryRestart(v)
 		}
 	}()
 
 	for i := 0; i < len(msgs); i++ {
 		nproc++
+		atomic.AddInt32(&p.mcount, -1)
 		msg := msgs[i]
 		if pill, ok := msg.Msg.(poisonPill); ok {
 			// If we need to gracefuly stop, we process all the messages
@@ -207,6 +213,10 @@ func (p *process) Send(_ *PID, msg any, sender *PID) {
 }
 func (p *process) Shutdown() {
 	p.cleanup(nil)
+}
+
+func (p *process) Count() int {
+	return p.inbox.Count() + int(atomic.LoadInt32(&p.mcount))
 }
 
 func cleanTrace(stack []byte) []byte {

--- a/cluster/cluster_test.go
+++ b/cluster/cluster_test.go
@@ -331,7 +331,7 @@ func TestGetActiveByID(t *testing.T) {
 
 	pid1 := c1.Activate("player", NewActivationConfig().WithID("1"))
 	pid2 := c2.Activate("player", NewActivationConfig().WithID("2"))
-	time.Sleep(time.Millisecond * 10)
+	time.Sleep(time.Millisecond * 200)
 
 	pid := c1.GetActiveByID("player/1")
 	assert.NotNil(t, pid)
@@ -364,7 +364,7 @@ func TestGetActiveByKind(t *testing.T) {
 	pid2 := c2.Activate("player", NewActivationConfig().WithID("2"))
 	c1.Activate("foo", NewActivationConfig().WithID("2"))
 	c1.Activate("bar", NewActivationConfig().WithID("2"))
-	time.Sleep(time.Millisecond * 10)
+	time.Sleep(time.Millisecond * 200)
 
 	pids := c1.GetActiveByKind("player")
 	assert.Len(t, pids, 2)

